bridge/tracker/rootchain.go

```diff
@@ -0,0 +1,236 @@
+	"github.com/umbracle/go-web3/abi"
+	client "github.com/umbracle/go-web3/jsonrpc"
+	"math/big"
+	"os"
+	"path/filepath"
+)
+
+const (
+	//	required block depth for fetching events on the rootchain
+	DefaultBlockConfirmations = 6
+
+	//	Ropsten testnet
+	//rootchainWS   = "wss://ropsten.infura.io/ws/v3/17eac086ff36442ebd43737400eb71ca"
+
+	//	Polygon Edge
+	rootchainWS = "ws://127.0.0.1:10002/ws"
+)
+
+//	contractABI is used to create query filter
+//	that matches events defined in a smart contract.
+type contractABI struct {
+	//	address of smart contract
+	address web3.Address
+
+	// signatures of events defined in smart contract
+	events []*abi.Event
+}
+
+//	eventIDs returns all the event signatures (IDs)
+//	defined in the smart contract.
+func (c *contractABI) eventIDs() (IDs []*web3.Hash) {
+	for _, ev := range c.events {
+		id := ev.ID()
+		IDs = append(IDs, &id)
+	}
+
+	return
+}
+
+//	loadABIs parses contracts from raw map.
+func loadABIs(abisRaw map[string][]string) (contracts []*contractABI) {
+	for address, events := range abisRaw {
+		//	set smart contract address
+		contract := &contractABI{
+			address: web3.HexToAddress(address),
+		}
+
+		//	set each event (defined in contract)
+		for _, ev := range events {
+			contract.events = append(contract.events, abi.MustNewEvent(ev))
+		}
+
+		//	append result
+		contracts = append(contracts, contract)
+	}
+
+	return
+}
+
+//	setupQueryFilter creates a log filter for the desired
+//	block range. Filter matches events defined in rootchain.go.
+func setupQueryFilter(from, to *big.Int, contracts []*contractABI) *web3.LogFilter {
+	queryFilter := &web3.LogFilter{}
+
+	//	set range of blocks to query
+	queryFilter.SetFromUint64(from.Uint64())
+	queryFilter.SetToUint64(to.Uint64())
+
+	//	set contract addresses and topics
+	for _, contract := range contracts {
+		//	append address
+		queryFilter.Address = append(queryFilter.Address, contract.address)
+
+		//	topics from all contracts must be in Topics[0]
+		queryFilter.Topics = append(queryFilter.Topics, contract.eventIDs())
+	}
+
+	return queryFilter
+}
+
+/* Header types parsed by the client */
+
+//	Ethereum header
+//type ethHeader struct {
+//	Difficulty   string        `json:"difficulty"`
+//	ExtraData    string        `json:"extraData"`
+//	GasLimit     string        `json:"gasLimit"`
+//	GasUsed      string        `json:"gasUsed"`
+//	LogsBloom    types.Bloom   `json:"logsBloom"`
+//	Miner        types.Address `json:"miner"`
+//	Nonce        string        `json:"nonce"`
+//	Number       string        `json:"number"`
+//	ParentHash   types.Hash    `json:"parentHash"`
+//	ReceiptsRoot types.Hash    `json:"receiptsRoot"`
+//	Sha3Uncles   types.Hash    `json:"sha3Uncles"`
+//	StateRoot    types.Hash    `json:"stateRoot"`
+//	Timestamp    string        `json:"timestamp"`
+//	TxRoot       types.Hash    `json:"transactionsRoot"`
+//	MixHash      types.Hash    `json:"mixHash"`
+//	Hash         types.Hash    `json:"hash"`
+//}
```

Comment:
@dbrajovic why do we need this commented out struct? i can't remember what was the conclusion to the things around this? on which occasions does the first vs the second structs work? were we able to notice the difference in a textual diff??

Reply:
Conclusion was that if you wanna use polygon-edge as rootchain (as we did for PoC) then the `ethHeader` needs to be different - our header field types are different. I don't think we reached any consensus on this.